### PR TITLE
Refactor benches; add `SymbolLang` benches

### PR DESCRIPTION
**Pull request by nmheim**
|                                 | egg-symbollang      | egg-customlang    | Metatheory         |
|:--------------------------------|:-------------------:|:-----------------:|:------------------:|
| egraph_addexpr                  | 0.696 ± 0.0056 ms   |                   | 4.66 ± 0.22 ms     |
| basic_maths_simpl2              | 0.288 ± 0.00014 s   | 3.58 ± 0.0026 ms  | 14.1 ± 1.4 ms      |
| prop_logic_freges_theorem       | 1.64 ± 0.0019 ms    | 1.11 ± 0.00071 ms | 2.12 ± 0.057 ms    |
| calc_logic_demorgan             | 12.6 ± 0.0071 ms    | 25 ± 0.018 μs     | 0.107 ± 0.0018 ms  |
| calc_logic_freges_theorem       | 4.17 ± 0.0018 ms    | 2.14 ± 0.00082 ms |                    |
| basic_maths_simpl1              | 0.293 ± 0.00013 s   | 2.13 ± 0.00075 ms | 6.77 ± 0.21 ms     |
| egraph_constructor              | 0.0542 ± 4.4e-05 μs |                   | 0.5 ± 0.084 μs     |
| prop_logic_prove1               | 20.2 ± 0.014 ms     | 19.1 ± 0.017 ms   | 0.0525 ± 0.0015 s  |
| prop_logic_demorgan             | 0.0559 ± 3.3e-05 ms | 31.3 ± 0.06 μs    | 0.137 ± 0.003 ms   |
| while_superinterpreter_while_10 |                     |                   | 20.7 ± 1.3 ms      |
| prop_logic_rewrite              |                     |                   | 0.0744 ± 0.0005 ms |
| time_to_load                    |                     |                   | 0.0427 ± 0.00046 s |

### CODE REVIEW

**Comment by nmheim** [comment]
Speedup tables

|                                 | egg-symbollang      | Metatheory         | egg-symbollang/Metatheory |
|:--------------------------------|:-------------------:|:------------------:|:-------------------------:|
| egraph_addexpr                  | 0.696 ± 0.0056 ms   | 4.66 ± 0.22 ms     | 0.149                     |
| basic_maths_simpl2              | 0.288 ± 0.00014 s   | 14.1 ± 1.4 ms      | 20.4                      |
| prop_logic_freges_theorem       | 1.64 ± 0.0019 ms    | 2.12 ± 0.057 ms    | 0.774                     |
| calc_logic_demorgan             | 12.6 ± 0.0071 ms    | 0.107 ± 0.0018 ms  | 118                       |
| calc_logic_freges_theorem       | 4.17 ± 0.0018 ms    |                    |                           |
| basic_maths_simpl1              | 0.293 ± 0.00013 s   | 6.77 ± 0.21 ms     | 43.3                      |
| egraph_constructor              | 0.0542 ± 4.4e-05 μs | 0.5 ± 0.084 μs     | 0.108                     |
| prop_logic_prove1               | 20.2 ± 0.014 ms     | 0.0525 ± 0.0015 s  | 0.385                     |
| prop_logic_demorgan             | 0.0559 ± 3.3e-05 ms | 0.137 ± 0.003 ms   | 0.408                     |
| while_superinterpreter_while_10 |                     | 20.7 ± 1.3 ms      |                           |
| prop_logic_rewrite              |                     | 0.0744 ± 0.0005 ms |                           |
| time_to_load                    |                     | 0.0427 ± 0.00046 s |                           |

|                                 | egg-customlang    | Metatheory         | egg-customlang/Metatheory |
|:--------------------------------|:-----------------:|:------------------:|:-------------------------:|
| prop_logic_freges_theorem       | 1.11 ± 0.00071 ms | 2.12 ± 0.057 ms    | 0.525                     |
| basic_maths_simpl2              | 3.58 ± 0.0026 ms  | 14.1 ± 1.4 ms      | 0.253                     |
| calc_logic_demorgan             | 25 ± 0.018 μs     | 0.107 ± 0.0018 ms  | 0.235                     |
| calc_logic_freges_theorem       | 2.14 ± 0.00082 ms |                    |                           |
| basic_maths_simpl1              | 2.13 ± 0.00075 ms | 6.77 ± 0.21 ms     | 0.315                     |
| prop_logic_prove1               | 19.1 ± 0.017 ms   | 0.0525 ± 0.0015 s  | 0.363                     |
| prop_logic_demorgan             | 31.3 ± 0.06 μs    | 0.137 ± 0.003 ms   | 0.229                     |
| egraph_addexpr                  |                   | 4.66 ± 0.22 ms     |                           |
| egraph_constructor              |                   | 0.5 ± 0.084 μs     |                           |
| while_superinterpreter_while_10 |                   | 20.7 ± 1.3 ms      |                           |
| prop_logic_rewrite              |                   | 0.0744 ± 0.0005 ms |                           |
| time_to_load                    |                   | 0.0427 ± 0.00046 s |                           |

File: benches/symbollang/basic_maths.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        |b| b.iter(|| {
+            simplify(black_box("(+ a (+ b (+ (* 0 c) d)))"));
+            //assert_eq!(result, "(+ d (+ b a))");
+        })
+    );
+
+    c.bench_function(
+        "basic_maths/simpl2",
+        |b| b.iter(|| {
+            let result =  simplify(black_box("(+ (+ (+ 0 (* (* 1 foo) 0)) (* a 0)) a)"));
```
Comment (0x0f0f0f) [inline benches/symbollang/basic_maths.rs:100]: Please use `prove` with same hyperparameters in places where `prove`  is used

File: benches/symbollang/calc_logic.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            assert!(tru.eq(&res))
+        })
+    );
+
+    let frege: RecExpr<SymbolLang> = "(=> (=> p (=> p r)) (=> (=> q p) (=> p r)))"
+        .parse().unwrap();
+    c.bench_function( "calc_logic/freges_theorem",
+        |b| b.iter(|| {
+            let res = simplify(black_box(&frege), black_box(&rules), 10);
+            assert!(tru.eq(&res))
```
Comment (0x0f0f0f) [inline benches/symbollang/calc_logic.rs:98]: same as above

**Comment by nmheim** [comment]
Fixed benchmark speedup tables

|                                 | egg-symbollang      | Metatheory         | egg-symbollang/Metatheory |
|:--------------------------------|:-------------------:|:------------------:|:-------------------------:|
| egraph_addexpr                  | 0.737 ± 0.0015 ms   | 5.56 ± 0.21 ms     | 0.133                     |
| basic_maths_simpl2              | 8.09 ± 0.027 ms     | 17.4 ± 1.4 ms      | 0.465                     |
| prop_logic_freges_theorem       | 1.7 ± 0.0019 ms     | 2.61 ± 0.061 ms    | 0.65                      |
| calc_logic_demorgan             | 0.0412 ± 3.4e-05 ms | 0.171 ± 0.0013 ms  | 0.241                     |
| calc_logic_freges_theorem       | 3.1 ± 0.0038 ms     |                    |                           |
| basic_maths_simpl1              | 4.15 ± 0.045 ms     | 8.54 ± 0.2 ms      | 0.486                     |
| egraph_constructor              | 0.0557 ± 3.1e-05 μs | 0.542 ± 0.12 μs    | 0.103                     |
| prop_logic_prove1               | 21 ± 0.037 ms       | 0.0614 ± 0.0019 s  | 0.342                     |
| prop_logic_demorgan             | 0.0574 ± 3e-05 ms   | 0.215 ± 0.0044 ms  | 0.267                     |
| while_superinterpreter_while_10 |                     | 25.8 ± 1.3 ms      |                           |
| prop_logic_rewrite              |                     | 0.0748 ± 0.0005 ms |                           |
| time_to_load                    |                     | 0.0433 ± 0.001 s   |                           |

|                                 | egg-customlang      | Metatheory         | egg-customlang/Metatheory |
|:--------------------------------|:-------------------:|:------------------:|:-------------------------:|
| egraph_addexpr                  | 0.737 ± 0.0015 ms   | 5.56 ± 0.21 ms     | 0.133                     |
| basic_maths_simpl2              | 8.09 ± 0.027 ms     | 17.4 ± 1.4 ms      | 0.465                     |
| prop_logic_freges_theorem       | 1.7 ± 0.0019 ms     | 2.61 ± 0.061 ms    | 0.65                      |
| calc_logic_demorgan             | 0.0412 ± 3.4e-05 ms | 0.171 ± 0.0013 ms  | 0.241                     |
| calc_logic_freges_theorem       | 3.1 ± 0.0038 ms     |                    |                           |
| basic_maths_simpl1              | 4.15 ± 0.045 ms     | 8.54 ± 0.2 ms      | 0.486                     |
| egraph_constructor              | 0.0557 ± 3.1e-05 μs | 0.542 ± 0.12 μs    | 0.103                     |
| prop_logic_prove1               | 21 ± 0.037 ms       | 0.0614 ± 0.0019 s  | 0.342                     |
| prop_logic_demorgan             | 0.0574 ± 3e-05 ms   | 0.215 ± 0.0044 ms  | 0.267                     |
| while_superinterpreter_while_10 |                     | 25.8 ± 1.3 ms      |                           |
| prop_logic_rewrite              |                     | 0.0748 ± 0.0005 ms |                           |
| time_to_load                    |                     | 0.0433 ± 0.001 s   |                           |

File: scripts/load_results.jl [hunk 3]
```diff
@@ -54,3 +54,16 @@ new_res = OrderedDict(
...
 AirspeedVelocity.create_table(new_res) |> print
+AirspeedVelocity.create_table(
+    OrderedDict(
+        "egg-symbollang" => new_res["egg-symbollang"],
+        "Metatheory" => new_res["Metatheory"],
+    )
+) |> print
+AirspeedVelocity.create_table(
+    OrderedDict(
+        "egg-customlang" => new_res["egg-symbollang"],
```
Comment (0x0f0f0f) [inline scripts/load_results.jl]:
```suggestion
        "egg-customlang" => new_res["egg-customlang"],
```

Comment (nmheim) [reply scripts/load_results.jl]: already fixed;) the table is with the correct key already:)

**Comment by nmheim** [comment]
Including benchmarks on `master`:

|                                 | egg-symbollang      | egg-customlang    | MT@3.0             | MT@2.0              |
|:--------------------------------|:-------------------:|:-----------------:|:------------------:|:-------------------:|
| egraph_addexpr                  | 0.737 ± 0.0015 ms   |                   | 5.56 ± 0.21 ms     | 10.9 ± 0.62 ms      |
| basic_maths_simpl2              | 8.09 ± 0.027 ms     | 3.49 ± 0.002 ms   | 17.4 ± 1.4 ms      | 0.787 ± 0.059 s     |
| prop_logic_freges_theorem       | 1.7 ± 0.0019 ms     | 1.08 ± 0.00099 ms | 2.61 ± 0.061 ms    | 30.9 ± 2.2 ms       |
| calc_logic_demorgan             | 0.0412 ± 3.4e-05 ms | 24.7 ± 0.03 μs    | 0.171 ± 0.0013 ms  | 0.328 ± 0.0071 ms   |
| calc_logic_freges_theorem       | 3.1 ± 0.0038 ms     | 2.12 ± 0.0029 ms  |                    |                     |
| basic_maths_simpl1              | 4.15 ± 0.045 ms     | 2.09 ± 0.0012 ms  | 8.54 ± 0.2 ms      | 0.0456 ± 0.0024 s   |
| egraph_constructor              | 0.0557 ± 3.1e-05 μs |                   | 0.542 ± 0.12 μs    | 0.416 ± 0.084 μs    |
| prop_logic_prove1               | 21 ± 0.037 ms       | 18.2 ± 0.018 ms   | 0.0614 ± 0.0019 s  | 7.3 s               |
| prop_logic_demorgan             | 0.0574 ± 3e-05 ms   | 29.7 ± 0.048 μs   | 0.215 ± 0.0044 ms  | 0.59 ± 0.011 ms     |
| while_superinterpreter_while_10 |                     |                   | 25.8 ± 1.3 ms      | 0.0737 ± 0.0023 s   |
| prop_logic_rewrite              |                     |                   | 0.0748 ± 0.0005 ms | 0.0803 ± 0.00079 ms |
| time_to_load                    |                     |                   | 0.0433 ± 0.001 s   | 0.0702 ± 0.0058 s   |

|                                 | egg-symbollang      | MT@3.0             | egg-symbollang/MT@3.0 |
|:--------------------------------|:-------------------:|:------------------:|:---------------------:|
| egraph_addexpr                  | 0.737 ± 0.0015 ms   | 5.56 ± 0.21 ms     | 0.133                 |
| basic_maths_simpl2              | 8.09 ± 0.027 ms     | 17.4 ± 1.4 ms      | 0.465                 |
| prop_logic_freges_theorem       | 1.7 ± 0.0019 ms     | 2.61 ± 0.061 ms    | 0.65                  |
| calc_logic_demorgan             | 0.0412 ± 3.4e-05 ms | 0.171 ± 0.0013 ms  | 0.241                 |
| calc_logic_freges_theorem       | 3.1 ± 0.0038 ms     |                    |                       |
| basic_maths_simpl1              | 4.15 ± 0.045 ms     | 8.54 ± 0.2 ms      | 0.486                 |
| egraph_constructor              | 0.0557 ± 3.1e-05 μs | 0.542 ± 0.12 μs    | 0.103                 |
| prop_logic_prove1               | 21 ± 0.037 ms       | 0.0614 ± 0.0019 s  | 0.342                 |
| prop_logic_demorgan             | 0.0574 ± 3e-05 ms   | 0.215 ± 0.0044 ms  | 0.267                 |
| while_superinterpreter_while_10 |                     | 25.8 ± 1.3 ms      |                       |
| prop_logic_rewrite              |                     | 0.0748 ± 0.0005 ms |                       |
| time_to_load                    |                     | 0.0433 ± 0.001 s   |                       |

|                                 | egg-customlang    | MT@3.0             | egg-customlang/MT@3.0 |
|:--------------------------------|:-----------------:|:------------------:|:---------------------:|
| prop_logic_freges_theorem       | 1.08 ± 0.00099 ms | 2.61 ± 0.061 ms    | 0.412                 |
| basic_maths_simpl2              | 3.49 ± 0.002 ms   | 17.4 ± 1.4 ms      | 0.201                 |
| calc_logic_demorgan             | 24.7 ± 0.03 μs    | 0.171 ± 0.0013 ms  | 0.145                 |
| calc_logic_freges_theorem       | 2.12 ± 0.0029 ms  |                    |                       |
| basic_maths_simpl1              | 2.09 ± 0.0012 ms  | 8.54 ± 0.2 ms      | 0.245                 |
| prop_logic_prove1               | 18.2 ± 0.018 ms   | 0.0614 ± 0.0019 s  | 0.297                 |
| prop_logic_demorgan             | 29.7 ± 0.048 μs   | 0.215 ± 0.0044 ms  | 0.138                 |
| egraph_addexpr                  |                   | 5.56 ± 0.21 ms     |                       |
| egraph_constructor              |                   | 0.542 ± 0.12 μs    |                       |
| while_superinterpreter_while_10 |                   | 25.8 ± 1.3 ms      |                       |
| prop_logic_rewrite              |                   | 0.0748 ± 0.0005 ms |                       |
| time_to_load                    |                   | 0.0433 ± 0.001 s   |                       |
|                                 | MT@2.0              | MT@3.0             | MT@2.0/MT@3.0 |
|:--------------------------------|:-------------------:|:------------------:|:-------------:|
| basic_maths_simpl1              | 0.0456 ± 0.0024 s   | 8.54 ± 0.2 ms      | 5.33          |
| basic_maths_simpl2              | 0.787 ± 0.059 s     | 17.4 ± 1.4 ms      | 45.3          |
| calc_logic_demorgan             | 0.328 ± 0.0071 ms   | 0.171 ± 0.0013 ms  | 1.92          |
| egraph_addexpr                  | 10.9 ± 0.62 ms      | 5.56 ± 0.21 ms     | 1.96          |
| egraph_constructor              | 0.416 ± 0.084 μs    | 0.542 ± 0.12 μs    | 0.768         |
| prop_logic_demorgan             | 0.59 ± 0.011 ms     | 0.215 ± 0.0044 ms  | 2.74          |
| prop_logic_freges_theorem       | 30.9 ± 2.2 ms       | 2.61 ± 0.061 ms    | 11.9          |
| prop_logic_prove1               | 7.3 s               | 0.0614 ± 0.0019 s  | 119           |
| prop_logic_rewrite              | 0.0803 ± 0.00079 ms | 0.0748 ± 0.0005 ms | 1.07          |
| while_superinterpreter_while_10 | 0.0737 ± 0.0023 s   | 25.8 ± 1.3 ms      | 2.85          |
| time_to_load                    | 0.0702 ± 0.0058 s   | 0.0433 ± 0.001 s   | 1.62          |

**Comment by nmheim** [comment]
|                                 | egg-sym              | egg-cust             | MT@3.0                 | MT@2.0                   | egg-sym/MT@3.0 | egg-cust/MT@3.... | MT@2.0/MT@3.0 |
|:--------------------------------|:--------------------:|:--------------------:|:----------------------:|:------------------------:|:--------------:|:-----------------:|:-------------:|
| egraph_addexpr                  | 737.009 ± 1.55 μs    |                      | 5560.646 ± 209.98 μs   | 10872.083 ± 615.08 μs    | 0.133          |                   | 1.96          |
| basic_maths_simpl2              | 8087.366 ± 27.10 μs  | 3490.326 ± 1.95 μs   | 17389.708 ± 1356.02 μs | 787362.458 ± 58976.50 μs | 0.465          | 0.201             | 45.3          |
| prop_logic_freges_theorem       | 1696.287 ± 1.86 μs   | 1076.140 ± 0.99 μs   | 2609.959 ± 61.21 μs    | 30949.458 ± 2186.29 μs   | 0.65           | 0.412             | 11.9          |
| calc_logic_demorgan             | 41.162 ± 0.03 μs     | 24.679 ± 0.03 μs     | 170.667 ± 1.29 μs      | 327.792 ± 7.08 μs        | 0.241          | 0.145             | 1.92          |
| calc_logic_freges_theorem       | 3098.192 ± 3.82 μs   | 2117.031 ± 2.89 μs   |                        |                          |                |                   |               |
| basic_maths_simpl1              | 4148.492 ± 44.81 μs  | 2088.562 ± 1.25 μs   | 8541.375 ± 201.54 μs   | 45550.375 ± 2390.25 μs   | 0.486          | 0.245             | 5.33          |
| egraph_constructor              | 0.056 ± 0.00 μs      |                      | 0.542 ± 0.12 μs        | 0.416 ± 0.08 μs          | 0.103          |                   | 0.768         |
| prop_logic_prove1               | 20983.521 ± 37.33 μs | 18236.646 ± 18.46 μs | 61431.334 ± 1932.70 μs | 7.3e+06 μs               | 0.342          | 0.297             | 119           |
| prop_logic_demorgan             | 57.371 ± 0.03 μs     | 29.701 ± 0.05 μs     | 215.083 ± 4.38 μs      | 589.583 ± 11.29 μs       | 0.267          | 0.138             | 2.74          |
| while_superinterpreter_while_10 |                      |                      | 25813.854 ± 1326.86 μs | 73677.292 ± 2340.71 μs   |                |                   | 2.85          |
| prop_logic_rewrite              |                      |                      | 74.834 ± 0.50 μs       | 80.250 ± 0.79 μs         |                |                   | 1.07          |
| time_to_load                    |                      |                      | 43268.791 ± 1025.46 μs | 70196.708 ± 5829.00 μs   |                |                   | 1.62          |

**Comment by nmheim** [comment]
|                                 | egg-sym   | egg-cust | MT@2.0     | MT@3.0   | egg-sym/MT@3.0 | egg-cust/MT@3.... | MT@2.0/MT@3.0 |
|:--------------------------------|:---------:|:--------:|:----------:|:--------:|:--------------:|:-----------------:|:-------------:|
| egraph_addexpr                  | 737 μs    |          | 10.9 ms    | 5.56 ms  | 0.133          |                   | 1.96          |
| basic_maths_simpl2              | 8.09 ms   | 3.49 ms  | 787 ms     | 17.4 ms  | 0.465          | 0.201             | 45.3          |
| prop_logic_freges_theorem       | 1.7 ms    | 1.08 ms  | 30.9 ms    | 2.61 ms  | 0.65           | 0.412             | 11.9          |
| calc_logic_demorgan             | 41.2 μs   | 24.7 μs  | 328 μs     | 171 μs   | 0.241          | 0.145             | 1.92          |
| calc_logic_freges_theorem       | 3.1 ms    | 2.12 ms  |            |          |                |                   |               |
| basic_maths_simpl1              | 4.15 ms   | 2.09 ms  | 45.6 ms    | 8.54 ms  | 0.486          | 0.245             | 5.33          |
| egraph_constructor              | 0.0557 μs |          | 0.416 μs   | 0.542 μs | 0.103          |                   | 0.768         |
| prop_logic_prove1               | 21 ms     | 18.2 ms  | 7.3e+03 ms | 61.4 ms  | 0.342          | 0.297             | 119           |
| prop_logic_demorgan             | 57.4 μs   | 29.7 μs  | 590 μs     | 215 μs   | 0.267          | 0.138             | 2.74          |
| while_superinterpreter_while_10 |           |          | 73.7 ms    | 25.8 ms  |                |                   | 2.85          |
| prop_logic_rewrite              |           |          | 80.2 μs    | 74.8 μs  |                |                   | 1.07          |
| time_to_load                    |           |          | 70.2 ms    | 43.3 ms  |                |                   | 1.62          |

**Pull request merged**